v3/src/components/graph/utilities/graph-utils.ts

```diff
@@ -198,49 +174,191 @@ export function percentString(value: number) {
   }).format(value)
 }
 
+function roundToSignificantDigits(value: number, sigDigits: number) {
+  if (isNaN(value) || value === 0) {
+    return { roundedValue: value, decPlaces: 0 }
+  }
+
+  const sign = value < 0 ? -1 : 1
+  value = Math.abs(value)
+
+  // Get the value in the range 10^(iSigDigits-1) to 10^iSigDigits.
+  const lower = Math.pow(10, sigDigits - 1)
+  const upper = lower * 10
+  let adjustedPlaces = 0
+  while (value > upper) {
+    value = value / 10
+    adjustedPlaces++
+  }
+  while (value < lower) {
+    value = value * 10
+    adjustedPlaces--
+  }
+
+  let newValue = Math.floor(value)
+  if (value - newValue > 0.5) {
+    newValue++
+  }
+
+  let counter = adjustedPlaces
+  while (counter < 0) {
+    newValue = newValue / 10
+    counter++
+  }
+
+  if (adjustedPlaces < 0) {
+    newValue = Number(newValue.toFixed(-adjustedPlaces))
+  }
+
+  while (counter > 0) {
+    newValue = newValue * 10
+    counter--
+  }
+
+  const roundedValue = sign * newValue
+  const decPlaces = -adjustedPlaces
+
+  return { roundedValue, decPlaces }
+}
+
+export function findNeededFractionDigits(slope: number, intercept: number, layout: GraphLayout) {
+  const xScale = layout.getAxisScale("bottom") as ScaleNumericBaseType
+  const yScale = layout.getAxisScale("left") as ScaleNumericBaseType
+  const xDomain = xScale.domain()
+  const yDomain = yScale.domain()
+  const kMaxDigits = 12  // Never try for more than this
+  let currCoords: IAxisIntercepts = {pt1: {x: 0, y: 0}, pt2: {x: 0, y: 0}}
+  let trialCoords: IAxisIntercepts = {pt1: {x: 0, y: 0}, pt2: {x: 0, y: 0}}
+  let tGreatestDiff = 0
+  let interceptDigits = 2
+  let slopeDigits = (slope < 0.001 && slope !== 0) ? Math.abs(Math.floor(Math.log10(Math.abs(slope)))) + 1 : 3
+
+  function convertToScreen(iWorldPts: IAxisIntercepts) {
+    return {
+      pt1: { x: xScale(iWorldPts.pt1.x), y: yScale(iWorldPts.pt1.y) },
+      pt2: { x: xScale(iWorldPts.pt2.x), y: yScale(iWorldPts.pt2.y) }
+    }
+  }
+
+  function greatestDiff() {
+    return Math.max(
+      Math.abs(trialCoords.pt1.x - currCoords.pt1.x),
+      Math.abs(trialCoords.pt1.y - currCoords.pt1.y),
+      Math.abs(trialCoords.pt2.x - currCoords.pt2.x),
+      Math.abs(trialCoords.pt2.y - currCoords.pt2.y)
+    )
+  }
+
+  if (isNaN(slope) || isNaN(intercept) || !slope || !intercept) {
```

Comment:
Slope and intercept can validly be 0. Is treating them the same as NaNs correct here? Could use `isFiniteNumber()` instead.

Reply:
@kswenson I believe you're right that using `isFiniteNumber()` would be better. This part is _almost_ verbatim from [the V2 version](https://github.com/concord-consortium/codap/blob/b8df5a48e552c8d6d19ff8a3d90d8a553e57ac9d/apps/dg/components/graph/utilities/plot_utilities.js#L301-L303). I think I was misinterpreting how `SC.none()` works -- `SC.none(0)` returns `false` instead of `true` like I wrongly assumed.

`if (isNaN(iSlope) || isNaN(iIntercept) || SC.none(iSlope) || SC.none(iIntercept)) {`

I've changed this in the latest commit. Does that look all right to you?

Reply:
`SC.none()` is essentially like `== null`, i.e. it captures `null` and `undefined`, so `isFiniteNumber()` should be an appropriate replacement.